coordinator/src/funding_fee.rs

```diff
@@ -0,0 +1,277 @@
+}
+
+#[derive(Clone, Copy, Debug, serde::Serialize, serde::Deserialize, PartialEq)]
+pub enum IndexPriceSource {
+    Bitmex,
+    Fake,
+}
+
+pub async fn generate_funding_fee_events_periodically(
+    scheduler: &JobScheduler,
+    pool: Pool<ConnectionManager<PgConnection>>,
+    schedule: String,
+    index_price_source: IndexPriceSource,
+) -> Result<()> {
+    scheduler
+        .add(tokio_cron_scheduler::Job::new(
+            schedule.as_str(),
+            move |_, _| {
+                if let Err(e) = generate_funding_fee_events(&pool, index_price_source) {
+                    tracing::error!("Failed to generate funding fee events: {e:#}");
+                }
+            },
+        )?)
+        .await?;
+
+    scheduler.start().await?;
+
+    Ok(())
+}
+
+/// Generate [`FundingFeeEvent`]s for all active positions.
+///
+/// When called, a [`FundingFeeEvent`] will be generated for an active position if:
+///
+/// - We can get a [`FundingRate`] that is at most 1 hour old from the DB.
+/// - We can get a BitMEX index price for the `end_date` of the [`FundingRate`].
+/// - There is no other [`FundingFeeEvent`] in the DB with the same `position_id` and `end_date`.
+/// - The position was created _before_ the `end_date` of the [`FundingRate`].
+fn generate_funding_fee_events(
+    pool: &Pool<ConnectionManager<PgConnection>>,
+    index_price_source: IndexPriceSource,
+) -> Result<()> {
+    let mut conn = pool.get()?;
+
+    let funding_rate = db::funding_rates::get_funding_rate_charged_in_the_last_hour(&mut conn)?;
+
+    let funding_rate = match funding_rate {
+        Some(funding_rate) => funding_rate,
+        None => {
+            tracing::debug!("No current funding rate for this hour");
+            return Ok(());
+        }
+    };
+
+    // TODO: Funding rates should be specific to contract symbols.
+    let contract_symbol = ContractSymbol::BtcUsd;
+
+    let current_index_price = match index_price_source {
+        IndexPriceSource::Bitmex => block_in_place(move || {
+            let current_index_price =
+                get_bitmex_index_price(&contract_symbol, funding_rate.end_date)?;
+
+            anyhow::Ok(current_index_price)
+        })?,
+        IndexPriceSource::Fake => {
+            #[cfg(not(debug_assertions))]
+            compile_error!("Cannot use a fake index price in release mode");
+
+            dec!(50_000)
+        }
+    };
+
+    if current_index_price.is_zero() {
+        bail!("Cannot generate funding fee events with zero index price");
+    }
+
+    // We exclude active positions which were open after this funding period ended.
+    let positions = db::positions::Position::get_all_active_positions_open_before(
+        &mut conn,
+        funding_rate.end_date,
+    )?;
+    for position in positions {
+        let amount =
+            calculate_funding_fee(position.quantity, funding_rate.rate, current_index_price);
+
+        if let Err(e) = db::funding_fee_events::insert(
```

Comment:
I mean, strictly speaking you could argue that the trader should get charged in this instant, and so if their margin falls below maintenance they should get liquidated.

But I'm not sure if I want to introduce the complexity of duplicating liquidation logic or sending this information to the liquidation task. Also, we would need to take into account all the other unpaid funding fee events.

Thinking about it again, the liquidation task could look for all the outstanding fees for a position and take that into account when deciding whether a position should get liquidated or not. That should be easy actually.

Reply:
> Thinking about it again, the liquidation task could look for all the outstanding fees for a position and take that into account when deciding whether a position should get liquidated or not. That should be easy actually.

Yes, this makes a lot of sense to me. Let's record this in a ticket please.

Reply:
I was going to try to address it in this PR, but I think it involves a big decision: should we update the `Position` after every new funding fee event is created? 

If so, it would be automatically taken into account when considering liquidation. But we currently expect the position to match the state of the DLC channel as much as possible, and that would no longer be the case if we do it this way.

If not, we would have to load all outstanding funding fee events for every position, every 30 seconds. That's very cumbersome and we would be recalculating the adapted liquidation price every 30 seconds too. Doesn't seem like a good idea.

I'll open the ticket like you suggest.

Reply:
https://github.com/get10101/10101/issues/2535